File: Commercial_Controller.go
package main

import (
	"fmt"
	"sort"
	"time"
)

// AppController Construct
type AppController struct {
	BatteryNum int
	arraybattery     []Battery
	columnNumber      int
	userInput      string
}


// Battery Construct
type Battery struct {
	columnNumber   int
	columnList []Column
}

// Column Construct
type Column struct {
	ColumnNumber  int
	elvPerColumn int
	ElevatorList  []Elevator
}

// Elevator Construct
type Elevator struct {
	elevatorNumber         int
	currentFloor      int
	floorList         []int
	state    		  string
	Direction 		  string
	doorstate         bool
	Column            Column
}

// NewController from Battery
func NewController(BatteryNum int) AppController {
	controller := new(AppController)
	controller.BatteryNum = 1
	for index := 0; index < BatteryNum; index++ {
		battery := NewBattery(index)
		controller.arraybattery = append(controller.arraybattery, *battery)
	}
	return *controller
}

// NewBattery to create column
func NewBattery(columnNumber int) *Battery {
	battery := new(Battery)
	battery.columnNumber = 4
	for index := 0; index < battery.columnNumber; index++ {
		Column := NewColumn(index)
		battery.columnList = append(battery.columnList, *Column)
	}
	return battery
}


// NewColumn to create Elevator
func NewColumn(elvPerColumn int) *Column {
	Column := new(Column)
	Column.elvPerColumn = 5
	for index := 0; index < Column.elvPerColumn; index++ {
		Elevator := NewElevator()
		Column.ElevatorList = append(Column.ElevatorList, *Elevator)
	}
	return Column
}

// NewElevator to define elevator
func NewElevator() *Elevator {
		Elevator := new(Elevator)
		Elevator.currentFloor = 7
		Elevator.floorList = []int{}
		Elevator.state = "idle"
		Elevator.Direction = "stop"
		Elevator.doorstate = true
		return Elevator
	}


// RequestElevator from userInput
func (controller *AppController) RequestElevator(FloorNumber, RequestedFloor int) Elevator {
	fmt.Println("Request Elevator to floor : ", FloorNumber)
	time.Sleep(2000 * time.Millisecond)
	var Column = controller.arraybattery[0].FindBestColumn(FloorNumber)
	controller.userInput = "down"
	var Elevator = Column.FindBestElevator(RequestedFloor, controller.userInput)
	Elevator.SendRequest(FloorNumber)
	Elevator.SendRequest(RequestedFloor)
	return Elevator
}

// AssignElevator from userInput
func (controller *AppController) AssignElevator(RequestedFloor int) Elevator {
	fmt.Println("Request Elevator to floor : ", RequestedFloor)
	time.Sleep(2000 * time.Millisecond)
	fmt.Println("Button Light On")
	Column := controller.arraybattery[0].FindBestColumn(RequestedFloor)
	controller.userInput = "up"
	var Elevator = Column.FindBestElevator(RequestedFloor, controller.userInput)
	var FloorNumber = 7
	Elevator.SendRequest(FloorNumber)
	Elevator.SendRequest(RequestedFloor)
	return Elevator
}

// FindBestColumn with ResquestFloor
func (b *Battery) FindBestColumn(RequestedFloor int) Column {
	if RequestedFloor >= 1 && RequestedFloor <= 7 { // B6(1) => B1(6) + RC(7)
		return b.columnList[0]
	} else if RequestedFloor > 8 && RequestedFloor <= 26 || RequestedFloor == 7 { //RC(7) + 2=>20=8=>26
		return b.columnList[1]
	} else if RequestedFloor > 27 && RequestedFloor <= 46 || RequestedFloor == 7 { //RC(7) + 21=>40=27=>46
		return b.columnList[2]
	} else if RequestedFloor > 47 && RequestedFloor <= 66 || RequestedFloor == 7 { //RC(7) + 41=>60=47=>60
		return b.columnList[3]
	}
	return b.columnList[3] //need to change 3 for ""
}

// FindBestElevator with RequestFloor, Direction
func (c *Column) FindBestElevator(RequestedFloor int, userInput string) Elevator {
	var selectedElevator = c.ElevatorList[4]
	for _, e := range c.ElevatorList {
		if RequestedFloor < e.currentFloor && e.Direction == "down" && userInput == "down" {
			selectedElevator = e
		} else if e.state == "idle" {
			selectedElevator = e
		} else if e.Direction != userInput && e.state == "moving" || e.state == "stopped" {
			selectedElevator = e
		} else if e.Direction == userInput && e.state == "moving" || e.state == "stopped" {
			selectedElevator = e
		}
	}
	return selectedElevator
}

// SendRequest to RequestFloor
func (e *Elevator) SendRequest(RequestedFloor int) {
	e.floorList = append(e.floorList, RequestedFloor)
	if RequestedFloor > e.currentFloor {

		sort.Ints(e.floorList)
	} else if RequestedFloor < e.currentFloor {

		sort.Sort(sort.Reverse(sort.IntSlice(e.floorList)))
	}
	e.OperateElevator(RequestedFloor)
}

//OperateElevator movement
func (e *Elevator) OperateElevator(RequestedFloor int) {
	if RequestedFloor == e.currentFloor {
		e.OpenDoor()
	} else if RequestedFloor > e.currentFloor {
		e.state = "moving"
		e.MoveUp(RequestedFloor)
		e.state = "stopped"
		e.OpenDoor()
		e.state = "moving"
	} else if RequestedFloor < e.currentFloor {
		e.state = "moving"
		e.MoveDown(RequestedFloor)
		e.state = "stopped"
		e.OpenDoor()
		e.state = "moving"
	}
}

// OpenDoor Elevator
func (e *Elevator) OpenDoor() {
	fmt.Println("Door is Opening")
	time.Sleep(1 * time.Second)
	fmt.Println("Door is Open")
	time.Sleep(1 * time.Second)
	e.CloseDoor()
}

// CloseDoor Elevator
func (e *Elevator) CloseDoor() {
	if e.doorstate == true {
		fmt.Println("Door is Closing")
		time.Sleep(1 * time.Second)
		fmt.Println("Door is Close")
		time.Sleep(1 * time.Second)
	} else if e.doorstate {
		e.OpenDoor()
		fmt.Println("Door can not be close please make sur door is not obstruct")
	}
}

// MoveUp Elevator
func (e *Elevator) MoveUp(RequestedFloor int) {
	fmt.Println("Column : ", + e.Column.ColumnNumber, " Elevator : #", + e.elevatorNumber, " Current Floor :", + e.currentFloor)
	for RequestedFloor > e.currentFloor {
		e.currentFloor ++
		if RequestedFloor == e.currentFloor {
			time.Sleep(1 * time.Second)
			fmt.Println("Column : ", + e.Column.ColumnNumber, " Elevator : #", + e.elevatorNumber, " Reach the destination floor : ", + e.currentFloor)
		}
		time.Sleep(500 * time.Millisecond)
		fmt.Println("Column : ", + e.Column.ColumnNumber, " Elevator : #", + e.elevatorNumber, " Floor : ", + e.currentFloor)
	}
}

// MoveDown Elevator
func (e *Elevator) MoveDown(RequestedFloor int) {
	fmt.Println("Column : ", + e.Column.ColumnNumber, " Elevator : #", + e.elevatorNumber, " Current Floor :", + e.currentFloor)
	for RequestedFloor < e.currentFloor {
		e.currentFloor --
		if RequestedFloor == e.currentFloor {
			time.Sleep(1 * time.Second)
			fmt.Println("Column : ", + e.Column.ColumnNumber, " Elevator : #", + e.elevatorNumber, " Reach the destination floor : ", + e.currentFloor)
		}
		time.Sleep(500 * time.Millisecond)
		fmt.Println("Column : ", + e.Column.ColumnNumber, " Elevator : #", + e.elevatorNumber, " Floor : ", + e.currentFloor)
	}
}

// main function in which test is performed \\
func main()  {
	controller := NewController(1)

	/* Scenario 4:
    With first column (or Column A) serving the basements B1 to B6, with elevator A1 idle at B4, A2 idle at 1st floor, A3 at B3 and going to B5, A4 at B6 and going to 1st floor, and A5 at B1 going to B6, someone is at B3 and requests the 1st floor. Elevator A4 is expected to be sent.
    */
	//column A
	
	// controller.arraybattery[0].columnList[0].ElevatorList[0].currentFloor = 3
	// controller.arraybattery[0].columnList[0].ElevatorList[0].state = "idle"
	// controller.arraybattery[0].columnList[0].ElevatorList[0].Direction = "stop"
	// controller.arraybattery[0].columnList[0].ElevatorList[0].SendRequest(3)
	
	// controller.arraybattery[0].columnList[0].ElevatorList[1].currentFloor = 7
	// controller.arraybattery[0].columnList[0].ElevatorList[1].state = "idle"
	// controller.arraybattery[0].columnList[0].ElevatorList[1].Direction = "stop"
	// controller.arraybattery[0].columnList[0].ElevatorList[1].SendRequest(7)
	
	// controller.arraybattery[0].columnList[0].ElevatorList[2].currentFloor = 4
	// controller.arraybattery[0].columnList[0].ElevatorList[2].state = "moving"
	// controller.arraybattery[0].columnList[0].ElevatorList[2].Direction = "down"
	// controller.arraybattery[0].columnList[0].ElevatorList[2].SendRequest(2)

	controller.arraybattery[0].columnList[0].ElevatorList[3].currentFloor = 1
	controller.arraybattery[0].columnList[0].ElevatorList[3].state = "moving"
	controller.arraybattery[0].columnList[0].ElevatorList[3].Direction = "up"
	controller.arraybattery[0].columnList[0].ElevatorList[3].SendRequest(7)
	
	// controller.arraybattery[0].columnList[0].ElevatorList[4].currentFloor = 6
	// controller.arraybattery[0].columnList[0].ElevatorList[4].state = "moving"
	// controller.arraybattery[0].columnList[0].ElevatorList[4].Direction = "up"
	// controller.arraybattery[0].columnList[0].ElevatorList[4].SendRequest(1)

	// controller.AssignElevator(4)
	controller.RequestElevator(4, 7)

	/* Scenario 1:
    With second column (or column B) serving floors from 2 to 20, with elevator B1 at 20th floor going to 5th, B2 at 3rd floor going to 15th, B3 at 13th floor going to 1st, B4 at 15th floor going to 2nd, and B5 at 6th floor going to 1st, someone is at 1st floor and requests the 20th floor, elevator B5 is expected to be sent
	*/

	// column B

	// controller.arraybattery[0].columnList[1].ElevatorList[0].currentFloor = 26
	// controller.arraybattery[0].columnList[1].ElevatorList[0].state = "moving"
	// controller.arraybattery[0].columnList[1].ElevatorList[0].Direction = "down"
	// controller.arraybattery[0].columnList[1].ElevatorList[0].SendRequest(11)

	// controller.arraybattery[0].columnList[1].ElevatorList[1].currentFloor = 9
	// controller.arraybattery[0].columnList[1].ElevatorList[1].state = "moving"
	// controller.arraybattery[0].columnList[1].ElevatorList[1].Direction = "up"
	// controller.arraybattery[0].columnList[1].ElevatorList[1].SendRequest(21)

	// controller.arraybattery[0].columnList[1].ElevatorList[2].currentFloor = 19
	// controller.arraybattery[0].columnList[1].ElevatorList[2].state = "moving"
	// controller.arraybattery[0].columnList[1].ElevatorList[2].Direction = "down"
	// controller.arraybattery[0].columnList[1].ElevatorList[2].SendRequest(7)

	// controller.arraybattery[0].columnList[1].ElevatorList[3].currentFloor = 21
	// controller.arraybattery[0].columnList[1].ElevatorList[3].state = "moving"
	// controller.arraybattery[0].columnList[1].ElevatorList[3].Direction = "down"
	// controller.arraybattery[0].columnList[1].ElevatorList[3].SendRequest(8)

	// controller.arraybattery[0].columnList[1].ElevatorList[4].currentFloor = 12
	// controller.arraybattery[0].columnList[1].ElevatorList[4].state = "moving"
	// controller.arraybattery[0].columnList[1].ElevatorList[4].Direction = "down"
	// controller.arraybattery[0].columnList[1].ElevatorList[4].SendRequest(7)

	// // controller.AssignElevator(26)
	// controller.RequestElevator(7, 26)

	/* Scenario 2:
    With third column (or column C) serving floors from 21 to 40, with elevator C1 at 1st floor going to 21th, C2 at 23st floor going to 28th, C3 at 33rd floor going to 1st, C4 at 40th floor going to 24th, and C5 at 39nd floor going to 1st, someone is at 1st floor and requests the 36th floor, elevator C1 is expected to be sent
	*/

	//column C

	// controller.arraybattery[0].columnList[2].ElevatorList[0].currentFloor = 7
	// controller.arraybattery[0].columnList[2].ElevatorList[0].state = "moving"
	// controller.arraybattery[0].columnList[2].ElevatorList[0].Direction = "up"
	// controller.arraybattery[0].columnList[2].ElevatorList[0].SendRequest(27)

	// controller.arraybattery[0].columnList[2].ElevatorList[1].currentFloor = 29
	// controller.arraybattery[0].columnList[2].ElevatorList[1].state = "moving"
	// controller.arraybattery[0].columnList[2].ElevatorList[1].Direction = "up"
	// controller.arraybattery[0].columnList[2].ElevatorList[1].SendRequest(34)

	// controller.arraybattery[0].columnList[2].ElevatorList[2].currentFloor = 39
	// controller.arraybattery[0].columnList[2].ElevatorList[2].state = "moving"
	// controller.arraybattery[0].columnList[2].ElevatorList[2].Direction = "down"
	// controller.arraybattery[0].columnList[2].ElevatorList[2].SendRequest(7)

	// controller.arraybattery[0].columnList[2].ElevatorList[3].currentFloor = 46
	// controller.arraybattery[0].columnList[2].ElevatorList[3].state = "moving"
	// controller.arraybattery[0].columnList[2].ElevatorList[3].Direction = "down"
	// controller.arraybattery[0].columnList[2].ElevatorList[3].SendRequest(30)

	// controller.arraybattery[0].columnList[2].ElevatorList[4].currentFloor = 45
	// controller.arraybattery[0].columnList[2].ElevatorList[4].state = "moving"
	// controller.arraybattery[0].columnList[2].ElevatorList[4].Direction = "down"
	// controller.arraybattery[0].columnList[2].ElevatorList[4].SendRequest(7

	// //controller.AssignElevator(27)
	// controller.RequestElevator(27, 42)

	/* Scenario 3:
    With fourth column (or column D) serving floors from 41 to 60, with elevator D1 at 58th floor going to 1st, D2 at 50th floor going to 60th, D3 at 46th floor going to 58th, D4 at 1st floor going to 54th, and D5 at 60th floor going to 1st, someone is at 54th floor and requests the 1st floor, elevator D1 is expected to pick him up
	*/

	//column D

	// controller.arraybattery[0].columnList[3].ElevatorList[0].currentFloor = 64
	// controller.arraybattery[0].columnList[3].ElevatorList[0].state = "moving"
	// controller.arraybattery[0].columnList[3].ElevatorList[0].Direction = "down"
	// controller.arraybattery[0].columnList[3].ElevatorList[0].SendRequest(7)

	// controller.arraybattery[0].columnList[3].ElevatorList[1].currentFloor = 56
	// controller.arraybattery[0].columnList[3].ElevatorList[1].state = "moving"
	// controller.arraybattery[0].columnList[3].ElevatorList[1].Direction = "up"
	// controller.arraybattery[0].columnList[3].ElevatorList[1].SendRequest(66)

	// controller.arraybattery[0].columnList[3].ElevatorList[2].currentFloor = 52
	// controller.arraybattery[0].columnList[3].ElevatorList[2].state = "moving"
	// controller.arraybattery[0].columnList[3].ElevatorList[2].Direction = "down"
	// controller.arraybattery[0].columnList[3].ElevatorList[2].SendRequest(64)

	// controller.arraybattery[0].columnList[3].ElevatorList[3].currentFloor = 7
	// controller.arraybattery[0].columnList[3].ElevatorList[3].state = "moving"
	// controller.arraybattery[0].columnList[3].ElevatorList[3].Direction = "down"
	// controller.arraybattery[0].columnList[3].ElevatorList[3].SendRequest(60)

	// controller.arraybattery[0].columnList[3].ElevatorList[4].currentFloor = 66
	// controller.arraybattery[0].columnList[3].ElevatorList[4].state = "moving"
	// controller.arraybattery[0].columnList[3].ElevatorList[4].Direction = "down"
	// controller.arraybattery[0].columnList[3].ElevatorList[4].SendRequest(7)

    // controller.AssignElevator(60)
    // controller.RequestElevator(60, 7)
}

File: Commercial_Controller.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace Commercial_controller
{

    public class AppController
    {
        public int nbFloor;
        public int nbEleInColumn;
        public int nbColumn;
        public string userMove;
        public Battery battery;
        public List<int> shortestList;

        public AppController(int nbFloor, int nbColumn, int nbEleInColumn, string userMove)
        {
            this.nbFloor = nbFloor;
            this.nbColumn = nbColumn;
            this.nbEleInColumn = nbEleInColumn;
            this.userMove = userMove;
            this.battery = new Battery(this.nbColumn);
        }

        //request elevator 
        public Elevator RequestElevator(int floorNumber, int requestedFloor)
        {
            Thread.Sleep(2000);
            Console.WriteLine("Request elevator to floor : " + floorNumber);
            Thread.Sleep(2000);
            var column = battery.findColumn(floorNumber);
            userMove = "down";
            var elevator = column.findRequestedEle(floorNumber, userMove);
            if (requestedFloor <= 0 && requestedFloor > 67)
            {
                Console.WriteLine("this floor is not vailde, try again if it fails again, maybe you're in the wrong building?");
            }
            else
            {
                if (elevator.currentFloor > floorNumber) //&& elevator.Direction == "down") 
                {
                    elevator.Request(floorNumber, column.columnNumber);
                    elevator.Request(requestedFloor, column.columnNumber);
                }
                else if (elevator.currentFloor < floorNumber) //&& elevator.Direction == "up")
                {
                    elevator.Move_down(requestedFloor, column.columnNumber);
                    elevator.Request(floorNumber, column.columnNumber);
                    elevator.Request(requestedFloor, column.columnNumber);
                }
                return elevator;

            }
            Console.WriteLine("Go back to idle");
            return elevator;
        }
        //door/movement
        public class Elevator
        {
            public int eleNumber;
            public string status;
            public int currentFloor;
            public string Direction;
            public bool sensor;
            public int floorDisplay;
            public List<int> floorList;

            public List<string> floorName;
            public Elevator(int eleNumber, string status, int currentFloor, string Direction)
            {
                this.eleNumber = eleNumber;
                this.status = status;
                this.currentFloor = currentFloor;
                this.Direction = Direction;
                this.floorDisplay = currentFloor;
                this.sensor = true;
                this.floorList = new List<int>();
                this.floorName = new List<string>();
            }

            // sendrequest receive information that user requested elevator and assign elevator
            public void Request(int requestedFloor, char columnNumber)
            {
                floorList.Add(requestedFloor);
                if (requestedFloor > currentFloor)
                {
                    floorList.Sort((a, b) => a.CompareTo(b));
                }
                else if (requestedFloor < currentFloor)
                {
                    floorList.Sort((a, b) => -1 * a.CompareTo(b));

                }

                checkEle(requestedFloor, columnNumber);
            }

            // task for 2 direction (up/down)
            public void checkEle(int requestedFloor, char columnNumber)
            {
                if (requestedFloor == currentFloor)
                {
                    Open_door();
                    this.status = "idle";

                    this.floorList.Remove(0);
                }
                else if (requestedFloor < this.currentFloor)
                {
                    status = "moving";
                    Console.WriteLine("Column : " + columnNumber + " Elevator : " + this.eleNumber + " " + status);
                    this.Direction = "down";
                    Move_down(requestedFloor, columnNumber);
                    this.status = "stopped";
                    Console.WriteLine("Column : " + columnNumber + " Elevator : " + this.eleNumber + " " + status);

                    this.Open_door();
                    this.floorList.Remove(0);
                }
                else if (requestedFloor > this.currentFloor)
                {
                    Thread.Sleep(2000);
                    this.status = "moving";
                    Console.WriteLine("Column : " + columnNumber + " Elevator : " + this.eleNumber + " " + status);
                    this.Direction = "up";
                    this.Move_up(requestedFloor, columnNumber);
                    this.status = "stopped";
                    Console.WriteLine("Column : " + columnNumber + " Elevator : " + this.eleNumber + " " + status);

                    this.Open_door();
                    this.floorList.Remove(0);
                }

            }
            // Open/Close Door
            public void Open_door()
            {
                Thread.Sleep(2000);

                Console.WriteLine("Door is Opening");

                Thread.Sleep(200);
                Console.WriteLine("Door is Open");
                Thread.Sleep(2000);

                this.Close_door();
            }
            public void Close_door()
            {
                if (sensor == true)
                {
                    Console.WriteLine("Door is Closing");
                    Thread.Sleep(200);
                    Console.WriteLine("Door is Close");
                    Thread.Sleep(200);
                }
                else if (sensor == false)
                {
                    Open_door();
                }
            }

            // moving direction

            public void Move_up(int requestedFloor, char columnNumber)
            {
                Console.WriteLine("Column : " + columnNumber + " Elevator : #" + eleNumber + "  Current Floor : " + this.currentFloor);
                Thread.Sleep(200);
                while (this.currentFloor != requestedFloor)
                {
                    this.currentFloor += 1;
                    Console.WriteLine("Column : " + columnNumber + " Elevator : #" + eleNumber + "  Floor : " + this.currentFloor);

                    Thread.Sleep(200);
                }
            }

            public void Move_down(int requestedFloor, char columnNumber)
            {
                Console.WriteLine("Column : " + columnNumber + " Elevator : #" + eleNumber + "  Current Floor : " + this.currentFloor);
                Thread.Sleep(200);

                while (this.currentFloor != requestedFloor)
                {
                    this.currentFloor -= 1;
                    Console.WriteLine("Column : " + columnNumber + " Elevator : #" + eleNumber + "  Floor : " + this.currentFloor);

                    Thread.Sleep(200);
                }

            }

        }

        // column redirect to best column create elevator

        public class Column
        {
            public char columnNumber;
            public int nbFloor;
            public int nbEleInColumn;
            public List<Elevator> eleList;
            public List<int> call_button_list;


            public Column(char columnNumber, int nbFloor, int nbEleInColumn)
            {
                this.columnNumber = columnNumber;
                this.nbFloor = nbFloor;
                this.nbEleInColumn = nbEleInColumn;
                eleList = new List<Elevator>();
                call_button_list = new List<int>();
                for (int i = 0; i < this.nbEleInColumn; i++)
                {
                    Elevator elevator = new Elevator(i, "idle", 7, "up");
                    eleList.Add(elevator);
                }
            }

            //find best elevator to user request to go up

            public Elevator findAssignElevator(int requestedFloor, int floorNumber, string userMove)
            {

                foreach (var elevator in eleList)
                    if (elevator.status == "idle")
                    {
                        return elevator;
                    }

                var bestElevator = 0;
                var shortest_distance = 999;
                for (var i = 0; i < this.eleList.Count; i++)
                {
                    var ref_distance = Math.Abs(eleList[i].currentFloor - eleList[i].floorList[0]) + Math.Abs(eleList[i].floorList[0] - 1);
                    if (shortest_distance >= ref_distance)
                    {
                        shortest_distance = ref_distance;
                        bestElevator = i;
                    }
                }
                return eleList[bestElevator];
            }

            //find best elevator to user request to go down

            public Elevator findRequestedEle(int requestedFloor, string userMove)
            {
                var shortest_distance = 999;
                var bestElevator = 0;

                for (var i = 0; i < this.eleList.Count; i++)
                {
                    var ref_distance = eleList[i].currentFloor - requestedFloor;

                    if (ref_distance > 0 && ref_distance < shortest_distance)
                    {
                        shortest_distance = ref_distance;
                        bestElevator = i;
                    }
                }
                return eleList[bestElevator];
            }

        }
        // battery create column  
        public class Battery
        {
            public string battery_status;
            public int nbColumn;
            public List<Column> column_list;


            public Battery(int nbColumn)
            {
                this.nbColumn = nbColumn;
                this.battery_status = "on";
                column_list = new List<Column>();



                char cols = 'A';
                for (int i = 0; i < this.nbColumn; i++, cols++)
                {

                    Column column = new Column(cols, 66, 5);

                    column.columnNumber = cols;
                    column_list.Add(column);

                }
            }
            // best column to find
            public Column findColumn(int requestedFloor)
            {
                Column best_column = null;
                foreach (Column column in column_list)
                {
                    if (requestedFloor > 1 && requestedFloor <= 6 || requestedFloor == 7)
                    {
                        best_column = column_list[0];
                    }
                    else if (requestedFloor > 8 && requestedFloor <= 27 || requestedFloor == 7)
                    {

                        best_column = column_list[1];


                    }
                    else if (requestedFloor > 28 && requestedFloor <= 47 || requestedFloor == 7)
                    {
                        best_column = column_list[2];


                    }
                    else if (requestedFloor > 48 && requestedFloor <= 66 || requestedFloor == 7)
                    {
                        best_column = column_list[3];


                    }
                }
                return best_column;
            }
        }
        //request user who want go up at floor X
        public Elevator assignElevator(int requestedFloor)
        {
            Thread.Sleep(2000);
            Console.WriteLine("Requested floor : " + requestedFloor);
            Thread.Sleep(2000);
            Console.WriteLine("Call Button Light On");


            Column column = battery.findColumn(requestedFloor);
            userMove = "up";
            var floorNumber = 7;
            Elevator elevator = column.findAssignElevator(requestedFloor, floorNumber, userMove);

            elevator.Request(floorNumber, column.columnNumber);

            elevator.Request(requestedFloor, column.columnNumber);

            return elevator;
        }




        public class CommercialCS
        {
            public static void Main(string[] args)
            {
                AppController controller = new AppController(66, 4, 5, "stop");


                // Test

                // Column A

                controller.battery.column_list[0].eleList[0].currentFloor = 3;
                controller.battery.column_list[0].eleList[0].Direction = "stop";
                controller.battery.column_list[0].eleList[0].status = "idle";
                //controller.battery.column_list[0].eleList[0].floorList.Add(3);


                controller.battery.column_list[0].eleList[1].currentFloor = 7;
                controller.battery.column_list[0].eleList[1].Direction = "stop";
                controller.battery.column_list[0].eleList[1].status = "idle";
                //controller.battery.column_list[0].eleList[1].floorList.Add(7);


                controller.battery.column_list[0].eleList[2].currentFloor = 4;
                controller.battery.column_list[0].eleList[2].Direction = "down";
                controller.battery.column_list[0].eleList[2].status = "moving";
                //controller.battery.column_list[0].eleList[2].floorList.Add(2);


                controller.battery.column_list[0].eleList[3].currentFloor = 1;
                controller.battery.column_list[0].eleList[3].Direction = "up";
                controller.battery.column_list[0].eleList[3].status = "moving";
                //controller.battery.column_list[0].eleList[3].floorList.Add(7);


                controller.battery.column_list[0].eleList[4].currentFloor = 6;
                controller.battery.column_list[0].eleList[4].Direction = "down";
                controller.battery.column_list[0].eleList[4].status = "moving";
                //controller.battery.column_list[0].eleList[4].floorList.Add(1);

                // controller.assignElevator(1);
                Elevator elevator = controller.RequestElevator(4, 7);

                // Column B

                // controller.battery.column_list[1].eleList[0].currentFloor = 25;
                // controller.battery.column_list[1].eleList[0].Direction = "down";
                // controller.battery.column_list[1].eleList[0].status = "moving";
                // controller.battery.column_list[1].eleList[0].floorList.Add(12);

                // controller.battery.column_list[1].eleList[1].currentFloor = 10;
                // controller.battery.column_list[1].eleList[1].Direction = "up";
                // controller.battery.column_list[1].eleList[1].status = "moving";
                // controller.battery.column_list[1].eleList[1].floorList.Add(13);

                // controller.battery.column_list[1].eleList[2].currentFloor = 20;
                // controller.battery.column_list[1].eleList[2].Direction = "down";
                // controller.battery.column_list[1].eleList[2].status = "moving";
                // controller.battery.column_list[1].eleList[2].floorList.Add(7);

                // controller.battery.column_list[1].eleList[3].currentFloor = 22;
                // controller.battery.column_list[1].eleList[3].Direction = "down";
                // controller.battery.column_list[1].eleList[3].status = "moving";
                // controller.battery.column_list[1].eleList[3].floorList.Add(9);


                // controller.battery.column_list[1].eleList[4].currentFloor = 13;
                // controller.battery.column_list[1].eleList[4].Direction = "down";
                // controller.battery.column_list[1].eleList[4].status = "moving";
                // controller.battery.column_list[1].eleList[4].floorList.Add(7);

                // controller.assignElevator(27);
                // //Elevator elevator = controller.RequestElevator(7, 27);

                // Column C

                // controller.battery.column_list[2].eleList[0].currentFloor = 7;
                // controller.battery.column_list[2].eleList[0].Direction = "up";
                // controller.battery.column_list[2].eleList[0].status = "moving";
                // controller.battery.column_list[2].eleList[0].floorList.Add(28);

                // controller.battery.column_list[2].eleList[1].currentFloor = 30;
                // controller.battery.column_list[2].eleList[1].Direction = "up";
                // controller.battery.column_list[2].eleList[1].status = "moving";
                // controller.battery.column_list[2].eleList[1].floorList.Add(35);

                // controller.battery.column_list[2].eleList[2].currentFloor = 47;
                // controller.battery.column_list[2].eleList[2].Direction = "down";
                // controller.battery.column_list[2].eleList[2].status = "moving";
                // controller.battery.column_list[2].eleList[2].floorList.Add(31);

                // controller.battery.column_list[2].eleList[3].currentFloor = 46;
                // controller.battery.column_list[2].eleList[3].Direction = "down";
                // controller.battery.column_list[2].eleList[3].status = "moving";
                // controller.battery.column_list[2].eleList[3].floorList.Add(7);

                // controller.battery.column_list[2].eleList[4].currentFloor = 47;
                // controller.battery.column_list[2].eleList[4].Direction = "down";
                // controller.battery.column_list[2].eleList[4].status = "moving";
                // controller.battery.column_list[2].eleList[4].floorList.Add(31);

                // controller.assignElevator(43);
                // // Elevator elevator = controller.RequestElevator(7, 43);


                // Column D

                // controller.battery.column_list[3].eleList[0].currentFloor = 64;
                // controller.battery.column_list[3].eleList[0].Direction = "down";
                // controller.battery.column_list[3].eleList[0].status = "moving";
                // controller.battery.column_list[3].eleList[0].floorList.Add(7);


                // controller.battery.column_list[3].eleList[1].currentFloor = 57;
                // controller.battery.column_list[3].eleList[1].Direction = "up";
                // controller.battery.column_list[3].eleList[1].status = "moving";
                // controller.battery.column_list[3].eleList[1].floorList.Add(66);


                // controller.battery.column_list[3].eleList[2].currentFloor = 53;
                // controller.battery.column_list[3].eleList[2].Direction = "up";
                // controller.battery.column_list[3].eleList[2].status = "moving";
                // controller.battery.column_list[3].eleList[2].floorList.Add(64);


                // controller.battery.column_list[3].eleList[3].currentFloor = 7;
                // controller.battery.column_list[3].eleList[3].Direction = "up";
                // controller.battery.column_list[3].eleList[3].status = "moving";
                // controller.battery.column_list[3].eleList[3].floorList.Add(60);


                // controller.battery.column_list[3].eleList[4].currentFloor = 66;
                // controller.battery.column_list[3].eleList[4].Direction = "down";
                // controller.battery.column_list[3].eleList[4].status = "moving";
                // controller.battery.column_list[3].eleList[4].floorList.Add(7);

                // // controller.assignElevator(54);
                // Elevator elevator = controller.RequestElevator(54, 7);

            }
        }
    }
}